lib/algos/cache/GapSet.ts

```diff
@@ -0,0 +1,352 @@
+import assert from 'assert';
+import SortedSet from 'collections/sorted-set';
+
+import errors from '../../errors';
+
+export type GapSetEntry = {
+    firstKey: string,
+    lastKey: string,
+    weight: number,
+};
+
+/**
+ * Specialized data structure to support caching of listing "gaps",
+ * i.e. ranges of keys that can be skipped over during listing
+ * (because they only contain delete markers as latest versions)
+ */
+export default class GapSet implements Iterable<GapSetEntry> {
+    _gaps: SortedSet;
+    _maxWeight: number;
+
+    /**
+     * @constructor
+
+     * @param {number} maxWeight - weight threshold for each cached
+     * gap (unitless). Triggers splitting gaps when reached
+     */
+    constructor(maxWeight: number) {
+        this._gaps = new SortedSet(
+            [],
+            (left, right) => left.firstKey === right.firstKey,
+            (left, right) => (
+                left.firstKey < right.firstKey ? -1 :
+                left.firstKey > right.firstKey ? 1 : 0
+            )
+        );
+        this._maxWeight = maxWeight;
+    }
+
+    /**
+     * Create a GapSet from an array of gap entries (used in tests)
+     */
+    static createFromArray(gaps: GapSetEntry[], maxWeight: number): GapSet {
+        const gapSet = new GapSet(maxWeight);
+        gapSet._gaps.addEach(gaps);
+        return gapSet;
+    }
+
+    /**
+     * Helper function for setGap() that initializes a new gap or
+     * retrieves an existing one depending on boundaries and weight
+     *
+     * @param {string} firstKey - first key of the gap
+     * @param {string} lastKey - last key of the gap
+     * @param {number} weight - total weight between 'firstKey' and 'lastKey'
+     * @return {object} - { gap: GapSetEntry, weightToMerge: number }
+     *   - gap: created or existing gap entry, to be possibly extended and merged
+     *   - weightToMerge: how much weight there is remaining to merge with the returned gap
+     */
+    _lookupOrCreateGap(firstKey: string, lastKey: string, weight: number): {
```

Comment:
> I have a bit of a hard time understanding this function's exact responsibility. Here's where I'm at: we're basically initializing the gap so that the firstKey is set (we actually don't care about it after this function call). Now we will either return an existing gap that already contains this key or a new gap that only contains this key. That choice depends on whether firstKey was already registered or not and whether we think the new gap will be too big to be merged with another one.

This is a good view of what the function does.